### PR TITLE
MM-21211: Optimize Channelstore.UpdateLastViewedAt

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

For postgres, we use CTEs to do both queries at one
and return the data.

For mysql, CTEs aren't available until version 8. So we just optimize the
existing query.

- We use IN instead of multiple OR conditions for the channel membership check.
According to mysql https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_in:

If no type conversion is needed for the values in the IN() list,
they are all constants of the same type, and expr can be compared to each of them
as a value of the same type (possibly after type conversion), an optimization takes place.

The values the list are sorted and the search for expr is done using a binary search,
which makes the IN() operation very quick.

- We use the previous column LastViewedAt to replace the value of LastUpdateAt
instead of doing a switch case again.

https://dev.mysql.com/doc/refman/5.6/en/ansi-diff-update.html

If you access a column from the table to be updated in an expression, UPDATE uses the
current value of the column.

Results (ran with getchannel profile):

**Postgres:**
#### POST /channels/members/me/view

| Metric | Baseline | Actual | Delta | Delta % |
| --- | --- | --- | --- | --- |
| Hits | 1594 | 1574 | -20 | -1.25%
| Error Rate | 0.00% | 0.00% | 0% | 0% |
| Mean Response Time | 7.08ms | 6.36ms | -0.72ms | -10.20% |
| Median Response Time | 5.00ms | 4.00ms | -1.00ms | -20.00% |
| 95th Percentile | 9.00ms | 8.00ms | -1.00ms | -11.11% |

**Mysql:**
#### POST /channels/members/me/view

| Metric | Baseline | Actual | Delta | Delta % |
| --- | --- | --- | --- | --- |
| Hits | 1600 | 1566 | -34 | -2.12%
| Error Rate | 0.00% | 0.00% | 0% | 0% |
| Mean Response Time | 7.39ms | 7.02ms | -0.37ms | -4.94% |
| Median Response Time | 6.00ms | 6.00ms | 0ms | 0% |
| 95th Percentile | 10.00ms | 10.00ms | 0ms | 0% |

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-21211

### CODE REVIEW

**Comment by Pomyk** [comment]
Did you test performance on channels with high user count? I see that the CTE version uses index idx_channelmembers_channel_id instead of primary key (at least on postgresql 9.4).

**Comment by agnivade** [comment]
Thanks for looking into it @Pomyk - Yes I have tested the performance using our loadtest which uses 1000 users which is the standard size of our installations.

From what I am seeing using postgres11, it is using the primary key:
```
 CTE Scan on c  (cost=34.01..34.05 rows=2 width=78) (actual time=0.039..0.039 rows=0 loops=1)
   CTE c
     ->  Bitmap Heap Scan on channels  (cost=8.58..15.65 rows=2 width=43) (actual time=0.037..0.038 rows=0 loops=1)
           Recheck Cond: ((id)::text = ANY ('{sfsg,25235f}'::text[]))
           ->  Bitmap Index Scan on channels_pkey  (cost=0.00..8.58 rows=2 width=0) (actual time=0.034..0.035 rows=0 loops=1)
                 Index Cond: ((id)::text = ANY ('{sfsg,25235f}'::text[]))
   CTE updated
     ->  Update on channelmembers cm  (cost=0.41..18.37 rows=1 width=319) (actual time=0.003..0.003 rows=0 loops=1)
           ->  Nested Loop  (cost=0.41..18.37 rows=1 width=319) (actual time=0.001..0.001 rows=0 loops=1)
                 ->  CTE Scan on c c_1  (cost=0.00..0.04 rows=2 width=196) (actual time=0.001..0.001 rows=0 loops=1)
                 ->  Index Scan using channelmembers_pkey on channelmembers cm  (cost=0.41..8.43 rows=1 width=193) (never executed)
                       Index Cond: (((channelid)::text = (c_1.id)::text) AND ((userid)::text = '2455'::text))
```

Are you saying you are not seeing the userId index being used at all in 9.4 ?

File: store/sqlstore/channel_store.go [hunk 1]
```diff
@@ -1652,41 +1652,62 @@ func (s SqlChannelStore) PermanentDeleteMembersByUser(userId string) *model.AppE
...
+	FROM c
+	WHERE cm.UserId = :UserId
+	AND c.Id=cm.ChannelId
+)
+SELECT Id, LastPostAt FROM c`
+	} else {
+		query = `SELECT Id, LastPostAt, TotalMsgCount FROM Channels WHERE Id IN ` + keys
+	}
+	_, err := s.GetMaster().Select(&lastPostAtTimes, query, props)
+	if err != nil || len(lastPostAtTimes) <= 0 {
```
[the remark this answers is not in the document]
Comment (agnivade) [reply store/sqlstore/channel_store.go]: I believe it checks for the case of no rows found, and hence adds the `extra = "No channels found"` statement. I think somehow the gorp library swallows the `sql.ErrNoRows` error.

[the remark this answers is not in the document]
Comment (streamer45) [reply store/sqlstore/channel_store.go]: Yes that part I understood. I was more interested in knowing why we also check for the length to be negative and not just equal to zero.

[the remark this answers is not in the document]
Comment (agnivade) [reply store/sqlstore/channel_store.go]: Oh _that_ part, sure, I agree. Will change.

**Comment by Pomyk** [comment]
Tested on 9.4 and 9.5 and the plan is:
```
 CTE Scan on c  (cost=16.73..16.75 rows=1 width=78) (actual time=0.051..0.052 rows=1 loops=1)
   CTE c
     ->  Index Scan using channels_pkey on channels  (cost=0.28..8.30 rows=1 width=43) (actual time=0.048..0.049 rows=1 loops=1)
           Index Cond: ((id)::text = 'rjemtnjdmpbeuj1ordr98ftbnh'::text)
   CTE updated
     ->  Update on channelmembers cm  (cost=0.29..8.44 rows=1 width=342) (actual time=0.376..0.376 rows=0 loops=1)
           ->  Nested Loop  (cost=0.29..8.44 rows=1 width=342) (actual time=0.060..0.178 rows=1 loops=1)
                 ->  CTE Scan on c c_1  (cost=0.00..0.02 rows=1 width=196) (actual time=0.008..0.008 rows=1 loops=1)
                 ->  Index Scan using idx_channelmembers_channel_id on channelmembers cm  (cost=0.29..8.41 rows=1 width=216) (actual time=0.049..0.166 rows=1 loops=1)
                       Index Cond: ((channelid)::text = (c_1.id)::text)
                       Filter: ((userid)::text = 'r44fm6ncmbn8uqeb4ozjxjd9ho'::text)
                       Rows Removed by Filter: 160
```

**Comment by Pomyk** [comment]
After analyze it started using primary key.

**Comment by agnivade** [comment]
There are already existing tests here - testChannelStoreUpdateLastViewedAt. Those test all the fields touched in the method to verify there are no regressions. Did you have any specific check in mind ?

**Comment by agnivade** [comment]
Hello @ogi-m - gentle reminder to take a look at this when you get a chance. Thanks !

**Pull request merged**